devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
+
+    def _extract_children_of_composites(self, kwargs):
+        new_params = {}
+        # If we've been passed CompositeFunction objects as kwargs,
+        # they might have children that need to be substituted as well.
+        for k, v in kwargs.items():
+            if isinstance(v, CompositeFunction):
+                orig_param_l = [i for i in self.parameters if i.name == k]
+                # If I have been passed a parameter, I must have seen it before
+                if len(orig_param_l) == 0:
+                    raise InvalidArgument("Parameter %s does not exist in expressions " +
+                                          "passed to this Operator" % k)
+                # We've made sure the list isn't empty. Names should be unique so it
+                # should have exactly one entry
+                assert(len(orig_param_l) == 1)
+                orig_param = orig_param_l[0]
+                # Pull out the children and add them to kwargs
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+        return kwargs
+
+    def _dle_arguments(self, dim_sizes):
+        # Add user-provided block sizes, if any
+        dle_arguments = OrderedDict()
+        autotune = True
+        for i in self.dle_arguments:
+            dim_size = dim_sizes.get(i.original_dim.name, None)
+            if dim_size is None:
+                error('Unable to derive size of dimension %s from defaults. '
+                      'Please provide an explicit value.' % i.original_dim.name)
+                raise InvalidArgument('Unknown dimension size')
+            if i.value:
+                try:
+                    dle_arguments[i.argument.name] = i.value(dim_size)
+                except TypeError:
+                    dle_arguments[i.argument.name] = i.value
+                    autotune = False
+            else:
+                dle_arguments[i.argument.name] = dim_size
+        return dle_arguments, autotune
+
+    def _derive_values(self, kwargs):
+        # Use kwargs
+        values = OrderedDict()
+        dimension_values = OrderedDict()
+        
+        for i in self.arguments:
+            values[i] = get_value(i, kwargs.pop(i.name, None), values)
+
+        for i in self.dims:
+            dimension_values[i] = kwargs.pop(i.name, None)
+
+        # Make sure we've used all arguments passed
+        if len(kwargs) > 0:
+            raise InvalidArgument("Unknown arguments passed: " + ", ".join(kwargs.keys()))
+
+        # Derive values for other arguments
+        for i in self.arguments:
+            if values[i] is None:
+                known_values = OrderedDict(chain(values.items(), dimension_values.items()))
+                provided_values = [get_value(i, x, known_values) for x in i.gets_value_from]
+                assert(len(provided_values) > 0)
+        
+                if len(provided_values) == 1:
```

Comment:
inline? or what happens if you just apply ``reduce``? there must be a way of dealing with length-one sequences

Reply:
[Explicit is better than implicit](https://www.python.org/dev/peps/pep-0020/)?

Reply:
OK so ``reduce`` expects at least two arguments, here you're providing one, how can it work?

And it should just work if you do:
``values[i] = reduce(function_u_want_to_apply, provided_values)``

As for the if-then-else question: I agree it's a matter of taste; what I do is that when they are (semantically) trivial, I try to write them as a single line. WHen I need to emphasise something (and perhaps I feel the need to add a comment), then I split it over multiple lines